### PR TITLE
#257 update graph schema

### LINKED ISSUE #257: Update graph schema
Discussion issue for discussing/documenting data + schema updates.

## Expected Behaviour

Entities:

```
Entity implements IEntity {
  uid: string;
  code: string;
  name: string;
  type: EEntityType | string;
  children: Entity[];
  properties: Property[];
}

CategoryEntity extends Entity implements ICategoryEntity {
  uid: string;
  code: string;
  name: string;
  type: EEntityType.Category;
  children: ProductEntity[];
  properties: CategoryProperty[];
}

ProductEntity extends Entity implements IProductEntity {
  uid: string;
  code: string;
  name: string;
  type: EEntityType.Product;
  combines: ProductEntity[];
  children: RouteEntity[];
  properties: ProductProperty[];
}

RouteEntity extends Entity implements IRouteEntity {
  uid: string;
  code: string;
  name: string;
  type: EEntityType.Route;
  children: FormEntity[];
  properties: RouteProperty[];
}

FormEntity extends Entity implements IFormEntity {
  uid: string;
  code: string;
  name: string;
  type: EEntityType.Form;
  children: (FormQualifierEntity | UnitEntity)[];
  properties: FormProperty[];
}

FormQualifierEntity extends Entity implements IFormQualifierEntity {
  uid: string;
  code: string;
  name: string;
  type: EEntityType.FormQualifier;
  children: UnitEntity[];
  properties: FormQualifierProperty[];
}

UnitEntity extends Entity implements IUnitEntity {
  uid: string;
  code: string;
  name: string;
  type: EEntityType.Unit;
  children: (PackImmediateEntity | PackSizeEntity)[];
  properties: UnitProperty[];
}

PackImmediateEntity extends Entity implements IPackImmediateEntity {
  uid: string;
  code: string;
  name: string;
  type: EEntityType.PackImmediate;
  children: PackSizeEntity[];
  properties: PackImmediateProperty[];
}

PackSizeEntity extends Entity implements IPackSizeEntity {
  uid: string;
  code: string;
  name: string;
  type: EEntityType.PackSize;
  children: (PackOuterEntity | ManufacturerEntity)[];
  properties: PackSizeProperty[];
}

PackOuterEntity extends Entity implements IPackOuterEntity {
  uid: string;
  code: string;
  name: string;
  type: EEntityType.PackOuter;
  children: ManufacturerEntity[];
  properties: PackOuterProperty[];
}

ManufacturerEntity extends Entity implements IManufacturerEntity {
  uid: string;
  code: string;
  name: string;
  type: EEntityType.Manufacturer;
  children: BrandEntity[];
  properties: ManufacturerProperty[];
}

BrandEntity extends Entity implements IBrandEntity {
  uid: string;
  code: string;
  name: string;
  type: EEntityType.Brand;
  children: undefined;
  properties: BrandProperty[];
}
```

Properties:

```
Property {
  type: PropertyType;
  value: object;
}

ProductInteractionProperty extends Property implements IProductInteractionProperty {
  type: PropertyType.ProductInteraction;
  value: {
    name: string;
    description: string;
    severity: string;
    source: string;
    rxcui: string;
  }
}
```

Other:

```
EntityResultSet {
  results: Entity[];
  size: number;
}

PropertyResultSet {
  results: Property[];
  size: number
}

ResultSet = EntityResultSet | PropertyResultSet;

EntityRequest {
  filter: EntityRequestFilter;
  first: number;
  offset: number;
}

EntityRequestFilter {
  code: string;
  name: string;
  orderBy: string;
  orderDesc: bool;
  type: string;
}
```

Notes:

- `Entity` split into explicit types for each level.
- `EntitySortInput` removed (my mistake!).
- `EntityCollection` updated to `EntityResultSet` and polymorphic `ResultSet` added.
- Redundant `Entity` specifiers removed.
- `Entity.description` renamed to `Entity.name`.
- Interactions to be resolved at `ProductEntity` level.
- Interactions to be given dedicated query, remove from entity and entities query.

## Current Behaviour

```
Entity {
  uid: string;
  code: string;
  description: string;
  type: string;
  interactions: DrugInteraction[];
  children: Entity[];
  properties: Property[];
}

EntityCollection {
  data: Entity[];
  totalLength: number;
}


EntitySearchInput {
  code: string;
  description: string;
  orderBy: EntitySort;
  type: string;
}

EntitySortInput {
  field: string;
  descending: boolean;
}

PropertyType {
  type: string;
  value: string;
  properties: Property[];
}

DrugInteractionType {
  name: string;
  description: string;
  severity: string;
  source: string;
  rxcui: string;
}
```

## Possible Solution

- Add `entity` types to `libs/data/src/lib/types/entity`.
- Add `property` types to `libs/data/src/lib/types/property`.
- Update query types + query logic.

### PR DESCRIPTION
**Pull request by wlthomson**
Fixes #257.

## Description

- Migrates existing graph to `v1/graphql`.
- Adds placeholder `v2/graphql`.
- Adds `v2` types.

### Documentation:

Tick below if one of the following applies:

1. Documentation has been added to cover changes introduced by this PR.
2. This PR requires no changes to documentation.

- [ ] Documentation OK.

### Tests:

Tick below if one of the following applies:

1. Tests have been added to cover changes introduced by this PR.
2. This PR requires no new tests.

- [ ] Tests OK.

### CODE REVIEW

File: unified-codes/apps/data-service/src/app/v2/types/entity/ProductEntity.ts [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+  readonly type: EEntityType.Product;
+  readonly combines?: IProductEntity[];
+  readonly children?: IProductEntityChildren;
+  readonly properties?: IProductEntityProperties;
+
+  constructor(product: IProductEntity) {
+    super(product);
+    this.type = EEntityType.Product;
+    this.combines = product.combines?.map((sibling: IProductEntity) => new ProductEntity(sibling));
+    this.children = product.children?.map((child: IProductEntityChild) => new RouteEntity(child));
```
Comment (katherine-sussol) [inline unified-codes/apps/data-service/src/app/v2/types/entity/ProductEntity.ts:37]: Just wondering about some products, as it seems some don't have a route defined (i.e. `ProductEntity` links straight to `DoseUnit`. One example of this is `Examination Glove`). Would we use an empty `RouteEntity` or allow product to have `DoseUnit` children?

File: unified-codes/apps/data-service/src/app/v2/types/property/Property.ts [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+  DrugInteraction = 'drug_interaction',
+}
+
+export interface IProperty {
+  type: EPropertyType;
+  // TODO: strongly type property values.
+  value: unknown;
+}
+
+// TODO: make Property abstract.
```
Comment (katherine-sussol) [inline unified-codes/apps/data-service/src/app/v2/types/property/Property.ts:16]: No changes required - just wondering what the plan is here? Was it to implement specific property types for all the possible properties? (e.g. NZULM, ATC etc).

**Review by katherine-sussol — APPROVED** [review]
Overall looks good to me, just a couple of minor questions. Not sure why the CI is marking it with a red cross either as it's building fine locally for me...

File: unified-codes/apps/data-service/src/app/v2/types/entity/ProductEntity.ts [hunk 3]
```diff
@@ -0,0 +1,42 @@
+import { EEntityType, IEntity, Entity } from './Entity';
+import { IRouteEntity, RouteEntity } from './RouteEntity';
+import { IProperty, Property } from '../Property/Property';
+
+export enum EProductField {
```
Comment (mark-prins) [inline unified-codes/apps/data-service/src/app/v2/types/entity/ProductEntity.ts:5]: the enum isn't used anywhere, can it be removed?

File: unified-codes/apps/data-service/src/app/v2/types/entity/FormEntity.ts [hunk 4]
```diff
@@ -0,0 +1,41 @@
+import { EEntityType, IEntity, Entity } from './Entity';
+import { IFormQualifierEntity, FormQualifierEntity } from './FormQualifierEntity';
+import { IUnitEntity, UnitEntity } from './UnitEntity';
+import { IProperty, Property } from '../Property';
+
+export enum EFormField {
```
Comment (mark-prins) [inline unified-codes/apps/data-service/src/app/v2/types/entity/FormEntity.ts:6]: The enum is unused, is it required?

File: unified-codes/apps/data-service/src/app/v2/types/property/DrugInteractionProperty.ts [hunk 5]
```diff
@@ -0,0 +1,45 @@
+import IProperty, { Property, EPropertyType } from './Property';
```
Comment (mark-prins) [inline unified-codes/apps/data-service/src/app/v2/types/property/DrugInteractionProperty.ts:1]: I don't think the `Property` in the name adds anything; `DrugInteraction` would be enough for me.

**Review by mark-prins — CHANGES_REQUESTED** [review]
the web project doesn't run - there's an issue importing `EEntityType` now that it has been removed from the export of data-service.

See `ExplorerReducer` which is doing this: `import { EEntityField, EEntityType } from '@unified-codes/data';
`

**Comment by mark-prins** [comment]
in fact there's a few places where types were imported by the web project. these will need to be redirected somehow

File: unified-codes/apps/data-service/src/app/v2/types/property/DrugInteractionProperty.ts (hunk 5 above)
Comment (wlthomson) [reply unified-codes/apps/data-service/src/app/v2/types/property/DrugInteractionProperty.ts:1]: It's just to start a convention, just like entities are all appended with `Entity` (just makes it clear that they inherit from the respective base class). Does that justify?

File: unified-codes/apps/data-service/src/app/v2/types/entity/ProductEntity.ts (hunk 3 above)
Comment (wlthomson) [reply unified-codes/apps/data-service/src/app/v2/types/entity/ProductEntity.ts:5]: Few things in `v2` which aren't currently used (as the resolvers and schema isn't implemented yet). Thinking these enums will probably be needed for the graph definitions (but if not, can remove them then!).

File: unified-codes/apps/data-service/src/app/v2/types/entity/FormEntity.ts (hunk 4 above)
Comment (wlthomson) [reply unified-codes/apps/data-service/src/app/v2/types/entity/FormEntity.ts:6]: Same as above 😨 !

**Comment by wlthomson** [comment]
> in fact there's a few places where types were imported by the web project. these will need to be redirected somehow

Argh, that's annoying. Guess the versioned types will have to be in the library (which tighly binds the library to the `data-service`)... can't really see another way of doing it!

File: unified-codes/apps/data-service/src/app/v2/types/property/Property.ts (hunk 2 above)
Comment (wlthomson) [reply unified-codes/apps/data-service/src/app/v2/types/property/Property.ts]: Yup, that's the idea!

noidea

**Comment by wlthomson** [comment]
@mark-prins @katherine-sussol Should be working properly now 🤞 . Once this is in can do the `v4.14` patch (hopefully still in time to avoid the wrath of @DhanyaHerath)!

File: unified-codes/apps/data-service/src/app/v2/types/property/DrugInteractionProperty.ts (hunk 5 above)
Comment (mark-prins) [reply unified-codes/apps/data-service/src/app/v2/types/property/DrugInteractionProperty.ts]: You're so conventional. let's run with it.

**Comment by mark-prins** [comment]
One final thing - the `.env` needs to be updated for the web to work. Would be good to have that in place `NX_DATA_SERVICE_GRAPHQL=v1/graphql` or a redirection from `/graphql` to the latest version...?

**Review by mark-prins — APPROVED** [review]
LGTM

**Pull request merged**